### DISCUSSION TITLE
Can geogram perform triangle re-meshing while retaining the features?

### DISCUSSION THREAD

**Discussion by mo1shang** [BrunoLevy/geogram#193] [Q&A]
Geogram claims to have fulfilled the function of the paper "Lp Centroidal Voronoi Tessellation and its applications", which claims to be able to achieve triangle re-meshization while maintaining sharp features. When I re-gridded the surface in geobox, I could not find the button related to the above paper, the default option resulted in a triangle mesh of good quality, but could not maintain the features.
I would like to know if geogram implements the thesis - Lp Centroidal Voronoi Tessellation and its applications, and if so, how can I use it.

**Comment by BrunoLevy** [comment]
Hello, 
No, "Lp Centroidal Voronoi Tessellation" is not part of geogram (and geogram does not "claim to have fullfilled the function of ...", why do you say that ?).
There is an implementation of "Lp Centroidal Voronoi Tessellation" that is part of Vorpaline,  marketed by the [Tessael](https://www.tessael.com/) company. Vorpaline has sharp feature preservation, quad and hex-dominant meshing capabilities. Do not hesitate to contact them if you are interested to acquire a license.

**Reply by mo1shang** [reply]
Thank you. It is a great honor to receive your reply. 
I didn't actually find an explicit statement that geogram implemented the thesis: Lp Centroidal Voronoi Tessellation and its applications, but I saw some statements on geogram's github home page and the links provided, and they went like this: “Geogram contains the main results in Geometry Processing from the former ALICE Inria project, that is, more than 30 research articles published in ACM SIGGRAPH, ACM Transactions on Graphics, Symposium on Geometry Processing and Eurographics...” . Among these 30 articles(https://github.com/BrunoLevy/geogram/wiki/Publications), one of them is Lp Centroidal Voronoi Tessellation and its applications. Therefore, I think geogram may have realized this paper. 
Since I am not a native English speaker, I may not have an accurate understanding of English. I am sorry for my dogmatism. 
Finally, I am actually concerned about the triangle re-meshing part of the above paper, if geogram implements this paper, it will be a great help to me, but it doesn't matter, geogram is still a powerful tool and helpful to me.

**Reply by BrunoLevy** [reply]
No problem :-) Thank you for the precision.

**Reply by mo1shang** [reply]
Thank you for your reply. To be honest, I was surprised to receive your reply so soon.
